sdk/program-error/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+    /// Allows on-chain programs to implement program-specific error types and see them returned
+    /// by the Solana runtime. A program-specific error may be any type that is represented as
+    /// or serialized to a u32 integer.
+    Custom(u32),
+    InvalidArgument,
+    InvalidInstructionData,
+    InvalidAccountData,
+    AccountDataTooSmall,
+    InsufficientFunds,
+    IncorrectProgramId,
+    MissingRequiredSignature,
+    AccountAlreadyInitialized,
+    UninitializedAccount,
+    NotEnoughAccountKeys,
+    AccountBorrowFailed,
+    MaxSeedLengthExceeded,
+    InvalidSeeds,
+    BorshIoError(String),
+    AccountNotRentExempt,
+    UnsupportedSysvar,
+    IllegalOwner,
+    MaxAccountsDataAllocationsExceeded,
+    InvalidRealloc,
+    MaxInstructionTraceLengthExceeded,
+    BuiltinProgramsMustConsumeComputeUnits,
+    InvalidAccountOwner,
+    ArithmeticOverflow,
+    Immutable,
+    IncorrectAuthority,
+}
+
+impl std::error::Error for ProgramError {}
+
+impl fmt::Display for ProgramError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ProgramError::Custom(num) => write!(f,"Custom program error: {num:#x}"),
+            ProgramError::InvalidArgument
+             => f.write_str("The arguments provided to a program instruction were invalid"),
+            ProgramError::InvalidInstructionData
+             => f.write_str("An instruction's data contents was invalid"),
+            ProgramError::InvalidAccountData
+             => f.write_str("An account's data contents was invalid"),
+            ProgramError::AccountDataTooSmall
+             => f.write_str("An account's data was too small"),
+            ProgramError::InsufficientFunds
+             => f.write_str("An account's balance was too small to complete the instruction"),
+            ProgramError::IncorrectProgramId
+             => f.write_str("The account did not have the expected program id"),
+            ProgramError::MissingRequiredSignature
+             => f.write_str("A signature was required but not found"),
+            ProgramError::AccountAlreadyInitialized
+             => f.write_str("An initialize instruction was sent to an account that has already been initialized"),
+            ProgramError::UninitializedAccount
+             => f.write_str("An attempt to operate on an account that hasn't been initialized"),
+            ProgramError::NotEnoughAccountKeys
+             => f.write_str("The instruction expected additional account keys"),
+            ProgramError::AccountBorrowFailed
+             => f.write_str("Failed to borrow a reference to account data, already borrowed"),
+            ProgramError::MaxSeedLengthExceeded
+             => f.write_str("Length of the seed is too long for address generation"),
+            ProgramError::InvalidSeeds
+             => f.write_str("Provided seeds do not result in a valid address"),
+            ProgramError::BorshIoError(s) =>  write!(f,"IO Error: {s}"),
```

Comment:
micro-nit: I'm surprised rustfmt didn't do anything about this
```suggestion
            ProgramError::BorshIoError(s) =>  write!(f, "IO Error: {s}"),
```